### ISSUE TITLE
Calling `OsuButton.TriggerClick` should animate and play sounds

### ISSUE THREAD

**Issue by peppy** [ppy/osu#14132]
Either that, or there should be a separate method which allows this. We are using `TriggerClick` in many places with this intention. This is especially important UX for keyboard navigation, where the user should be audibly and visually cued as to which action was performed by their keyboard input.

**Comment by Joehuu** [comment]
This is already possible.

For sounds, `TriggerClick` doesn't propagate per child's `OnClick` (reason why the `HoverClickSounds` logic doesn't work).

As for animation/flash, `OsuButton` simply doesn't have flashing on click anymore / regressed since https://github.com/ppy/osu/pull/6718 and https://github.com/ppy/osu-framework/pull/2913.

**Comment by peppy** [comment]
Not sure what you mean by "already possible". Your next line says that it isn't, right? That's why this issue was opened -- it should be triggering the `HoverClickSounds`.

**Comment by peppy** [comment]
Looking into this, `OsuAnimatedButton` already has more correct behaviour for the majority case. I think `OsuButton` can likely be combined with it, or at very least they should be sharing a base class.

I might look into this if no one else beats me.

**Comment by Joehuu** [comment]
Oh, I meant it's already possible with animations. As for the sounds, I would add a `PlayClickSample` method in `HoverClickSounds` and let the parent handle it on their `OnClick` instead but might be messy.

The button structure surely needs to be reworked. Is there a reason why `OsuButton` doesn't inherit `BasicButton` from framework?

**Comment by peppy** [comment]
`BasicButton` isn't made to be inherited (it's just meant as an initial implementation one can use until they want to stylise the component themselves). Is there a specific reason to use `BasicButton` here?

**Comment by Joehuu** [comment]
Not specific to this issue, but `BasicButton` already has the properties of what a button should have and is customizable. `Button` is [basically a `ClickableContainer`](https://github.com/ppy/osu-framework/blob/master/osu.Framework/Graphics/UserInterface/Button.cs).

`OsuTextBox` inherits `BasicTextBox`, so I don't get the structure `Button`/`BasicButton` is right now.

**Comment by peppy** [comment]
I think the eventual plan was to avoid the textbox inheritance too, but I'd need to dig up discussion there. That's one exception where the base class still needs a serious rewrite.

**Comment by frenzibyte** [comment]
I think it's more expected for `TriggerClick` to propagate an `OnClick` to a local input queue with the target drawable being the root, I was confused when I read the title why `TriggerClick` wouldn't play sounds, until I realized it invokes `OnClick` on just the target drawable and none of its children.

Also, I haven't realized that `OsuTextBox` inherits from `BasicTextBox`, but yeah the base class needs a full rewrite anyways.

**Comment by peppy** [comment]
That seems like it'd make sense for this specific use case, but not sure if we want to do that (feels like it might not fit with the hierarchy design). @smoogipoo thoughts?

**Comment by smoogipoo** [comment]
Yeah this is why I originally mentioned that `Click()` should be removed, though admittedly I was only looking at the single o!f use case at the time and not the bajillion others in osu! itself...

I have no immediate solutions to this problem and I'm not against the proposed solutions, so the following is just a few questions/general inquiries I have rather than immediate blockers.

I'm not sure if it feels right to propagate clicks to the inner hierarchy, because where exactly do you click? Do you click the drawable itself, or do you click a box in its upper left corner, or a box in its upper right corner? Nested handling seems to get fuzzy, since `OnClick` is a positional event. For example, what happens when someone does `OsuGame.TriggerClick()`?  
Fwiw, even the way it's currently done can lend itself to weirdness, like if for whatever reason you're accessing the mouse position in the click event then what would you expect that value to be?  
Is this being handled correctly in current cases under the absence of hover states (we usually have `updateHoverState()` methods to change colours and stuff, along with click-specific transforms to follow up)?

If I were to entertain the idea of nested propagation (and again I'm not against it, just so we're clear), I'm not sure how easy it is to get the input manager to actually only propagate input to a given hierarchy. Given that `ButtonEventManager`s store the currently pressed drawables I'd be worried about mangling this state. It'll require pretty big framework changes at least, I believe.

Thinking further, what about joystick handling? The ability to control the game without a pointing device is something we've talked about having. Is this functionality exclusive to `TriggerClick()` or should we also consider the possibility of `TriggerHover()`? If so, then maybe instead a more `ManualInputManager` approach is better with, say, a `ClickAndReturn()` method that does:
```
var lastPosition = InputManager.CurrentState.Mouse.Position;
InputManager.MoveMouseTo(drawable);
InputManager.Click(MouseButton.Left);
InputManager.MoveMouseTo(lastPosition);
```

**Comment by frenzibyte** [comment]
> If I were to entertain the idea of nested propagation (and again I'm not against it, just so we're clear), I'm not sure how easy it is to get the input manager to actually only propagate input to a given hierarchy. Given that ButtonEventManagers store the currently pressed drawables I'd be worried about mangling this state. It'll require pretty big framework changes at least, I believe.

I was thinking more of an isolated local logic in `CompositeDrawable`, like the following:
```csharp
         public override bool TriggerClick()
         {
             var state = GetContainingInputManager()?.CurrentState ?? new InputState();
             var click = new ClickEvent(state, MouseButton.Left);
             var queue = new List<Drawable>();

             BuildPositionalInputQueue(null, queue);
             queue.Reverse();

             var handledBy = queue.FirstOrDefault(target => target.TriggerEvent(click));

             if (handledBy == null)
                 return false;

             Logger.Log($"{GetType().ReadableName()}.{nameof(TriggerClick)}() handled by {handledBy}.", LoggingTarget.Runtime, LogLevel.Debug);
             return true;
         }
```

The point about `OsuGame` is a bit iffy since you wouldn't actually ever think of triggering that, but still yeah, I've "solved" that by actually never considering `ReceivePositionalInputAt` from this path, can also change it so that `ReceivePositionalInputAt` accepts a nullable vector and considers null as the "any position" value.

> Thinking further, what about joystick handling? The ability to control the game without a pointing device is something we've talked about having. Is this functionality exclusive to `TriggerClick()` or should we also consider the possibility of `TriggerHover()`? If so, then maybe instead a more `ManualInputManager` approach is better with, say, a `ClickAndReturn()` method that does:
> 
> ```
> var lastPosition = InputManager.CurrentState.Mouse.Position;
> InputManager.MoveMouseTo(drawable);
> InputManager.Click(MouseButton.Left);
> InputManager.MoveMouseTo(lastPosition);
> ```

How would usages be able to perform the click, DI'ing this manual input manager and requesting to click the target drawable?

**Comment by smoogipoo** [comment]
Yeah, I'm not against that implementation.

> How would usages be able to perform the click? DI'ing this manual input manager and requesting to click the target drawable?

Something like that, yeah. Inline `Dependencies.Get<SomeInputManager>().ClickAndReturn()` since it's not much more expensive than `GetContainingInputManager()`.

**Comment by peppy** [comment]
I believe this has been resolved.

**Comment by frenzibyte** [comment]
`OsuButton.TriggerClick` is still not playing sounds, keeping this open to track as most of the discussion here was with the intent of making sounds play.
